// @mysten/enoki/dist/cjs/wallet/state.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), setter ? setter.call(obj, value) : member.set(obj, value), value);
var __privateMethod = (obj, member, method) => (__accessCheck(obj, member, "access private method"), method);
var state_exports = {};
__export(state_exports, {
  INTERNAL_ONLY_EnokiFlow: () => INTERNAL_ONLY_EnokiFlow
});
module.exports = __toCommonJS(state_exports);
var import_webcrypto = require("@mysten/signers/webcrypto");
var import_zklogin = require("@mysten/sui/zklogin");
var import_idb_keyval = require("idb-keyval");
var import_nanostores = require("nanostores");
var import_encryption = require("../encryption.js");
var import_EnokiClient = require("../EnokiClient/index.js");
var import_EnokiKeypair = require("../EnokiKeypair.js");
var import_stores = require("../stores.js");
var _storageKeys, _enokiClient, _network, _encryption, _encryptionKey, _store, _idbStore, _INTERNAL_ONLY_EnokiFlow_instances, setSession_fn;
const createStorageKeys = (apiKey, network) => ({
  STATE: `@enoki/flow/state/${apiKey}/${network}`,
  SESSION: `@enoki/flow/session/${apiKey}/${network}`
});
class INTERNAL_ONLY_EnokiFlow {
  constructor(config) {
    __privateAdd(this, _INTERNAL_ONLY_EnokiFlow_instances);
    __privateAdd(this, _storageKeys);
    __privateAdd(this, _enokiClient);
    __privateAdd(this, _network);
    __privateAdd(this, _encryption);
    __privateAdd(this, _encryptionKey);
    __privateAdd(this, _store);
    __privateAdd(this, _idbStore);
    __privateSet(this, _enokiClient, new import_EnokiClient.EnokiClient({
      apiKey: config.apiKey,
      apiUrl: config.apiUrl
    }));
    __privateSet(this, _network, config.network);
    __privateSet(this, _encryptionKey, config.apiKey);
    __privateSet(this, _encryption, (0, import_encryption.createDefaultEncryption)());
    __privateSet(this, _store, (0, import_stores.createSessionStorage)());
    __privateSet(this, _storageKeys, createStorageKeys(config.apiKey, __privateGet(this, _network)));
    __privateSet(this, _idbStore, (0, import_idb_keyval.createStore)("enoki", `${config.apiKey}_${__privateGet(this, _network)}`));
    let storedState = null;
    try {
      const rawStoredValue = __privateGet(this, _store).get(__privateGet(this, _storageKeys).STATE);
      if (rawStoredValue) {
        storedState = JSON.parse(rawStoredValue);
      }
    } catch {
    }
    this.$zkLoginState = (0, import_nanostores.atom)(storedState || {});
    this.$zkLoginSession = (0, import_nanostores.atom)({ initialized: false, value: null });
    (0, import_nanostores.onMount)(this.$zkLoginSession, () => {
      this.getSession();
    });
    (0, import_nanostores.onSet)(this.$zkLoginState, ({ newValue }) => {
      __privateGet(this, _store).set(__privateGet(this, _storageKeys).STATE, JSON.stringify(newValue));
    });
  }
  get network() {
    return __privateGet(this, _network);
  }
  async createAuthorizationURL(input) {
    const ephemeralKeyPair = await import_webcrypto.WebCryptoSigner.generate();
    const { nonce, randomness, maxEpoch, estimatedExpiration } = await __privateGet(this, _enokiClient).createZkLoginNonce({
      network: __privateGet(this, _network),
      ephemeralPublicKey: ephemeralKeyPair.getPublicKey()
    });
    const params = new URLSearchParams({
      ...input.extraParams,
      nonce,
      client_id: input.clientId,
      redirect_uri: input.redirectUrl,
      response_type: "id_token",
      // TODO: Eventually fetch the scopes for this client ID from the Enoki service:
      scope: [
        "openid",
        // Merge the requested scopes in with the required openid scopes:
        ...input.extraParams && "scope" in input.extraParams ? input.extraParams.scope : []
      ].filter(Boolean).join(" ")
    });
    let oauthUrl;
    switch (input.provider) {
      case "google": {
        oauthUrl = `https://accounts.google.com/o/oauth2/v2/auth?${params}`;
        break;
      }
      case "facebook": {
        oauthUrl = `https://www.facebook.com/v17.0/dialog/oauth?${params}`;
        break;
      }
      case "twitch": {
        params.set("force_verify", "true");
        oauthUrl = `https://id.twitch.tv/oauth2/authorize?${params}`;
        break;
      }
      default:
        throw new Error(`Invalid provider: ${input.provider}`);
    }
    this.$zkLoginState.set({ provider: input.provider });
    await (0, import_idb_keyval.set)("ephemeralKeyPair", ephemeralKeyPair.export(), __privateGet(this, _idbStore));
    await __privateMethod(this, _INTERNAL_ONLY_EnokiFlow_instances, setSession_fn).call(this, {
      expiresAt: estimatedExpiration,
      maxEpoch,
      randomness
    });
    return oauthUrl;
  }
  async handleAuthCallback(hash = window.location.hash) {
    const params = new URLSearchParams(hash.startsWith("#") ? hash.slice(1) : hash);
    const zkp = await this.getSession();
    if (!zkp || !zkp.maxEpoch || !zkp.randomness) {
      throw new Error(
        "Start of sign-in flow could not be found. Ensure you have started the sign-in flow before calling this."
      );
    }
    const jwt = params.get("id_token");
    if (!jwt) {
      throw new Error("Missing ID Token");
    }
    (0, import_zklogin.decodeJwt)(jwt);
    const { address, salt, publicKey } = await __privateGet(this, _enokiClient).getZkLogin({ jwt });
    this.$zkLoginState.set({
      ...this.$zkLoginState.get(),
      salt,
      address,
      publicKey
    });
    await __privateMethod(this, _INTERNAL_ONLY_EnokiFlow_instances, setSession_fn).call(this, {
      ...zkp,
      jwt
    });
    return params.get("state");
  }
  async getSession() {
    if (this.$zkLoginSession.get().initialized) {
      return this.$zkLoginSession.get().value;
    }
    try {
      const storedValue = __privateGet(this, _store).get(__privateGet(this, _storageKeys).SESSION);
      if (!storedValue) return null;
      const state = JSON.parse(
        await __privateGet(this, _encryption).decrypt(__privateGet(this, _encryptionKey), storedValue)
      );
      if (state?.expiresAt && Date.now() > state.expiresAt) {
        await this.logout();
      } else {
        this.$zkLoginSession.set({ initialized: true, value: state });
      }
    } catch {
      this.$zkLoginSession.set({ initialized: true, value: null });
    }
    return this.$zkLoginSession.get().value;
  }
  async logout() {
    this.$zkLoginState.set({});
    __privateGet(this, _store).delete(__privateGet(this, _storageKeys).STATE);
    await (0, import_idb_keyval.clear)(__privateGet(this, _idbStore));
    await __privateMethod(this, _INTERNAL_ONLY_EnokiFlow_instances, setSession_fn).call(this, null);
  }
  // TODO: Should this return the proof if it already exists?
  async getProof() {
    const zkp = await this.getSession();
    const { salt } = this.$zkLoginState.get();
    if (zkp?.proof) {
      if (zkp.expiresAt && Date.now() > zkp.expiresAt) {
        throw new Error("Stored proof is expired.");
      }
      return zkp.proof;
    }
    if (!salt || !zkp || !zkp.jwt) {
      throw new Error("Missing required parameters for proof generation");
    }
    const storedNativeSigner = await (0, import_idb_keyval.get)(
      "ephemeralKeyPair",
      __privateGet(this, _idbStore)
    );
    if (!storedNativeSigner) {
      throw new Error("Native signer not found in store.");
    }
    const ephemeralKeyPair = import_webcrypto.WebCryptoSigner.import(storedNativeSigner);
    const proof = await __privateGet(this, _enokiClient).createZkLoginZkp({
      network: __privateGet(this, _network),
      jwt: zkp.jwt,
      maxEpoch: zkp.maxEpoch,
      randomness: zkp.randomness,
      ephemeralPublicKey: ephemeralKeyPair.getPublicKey()
    });
    await __privateMethod(this, _INTERNAL_ONLY_EnokiFlow_instances, setSession_fn).call(this, {
      ...zkp,
      proof
    });
    return proof;
  }
  async getKeypair() {
    await this.getProof();
    const zkp = await this.getSession();
    const { address } = this.$zkLoginState.get();
    if (!address || !zkp || !zkp.proof) {
      throw new Error("Missing required data for keypair generation.");
    }
    if (Date.now() > zkp.expiresAt) {
      throw new Error("Stored proof is expired.");
    }
    const storedNativeSigner = await (0, import_idb_keyval.get)(
      "ephemeralKeyPair",
      __privateGet(this, _idbStore)
    );
    if (!storedNativeSigner) {
      throw new Error("Native signer not found in store.");
    }
    const ephemeralKeypair = import_webcrypto.WebCryptoSigner.import(storedNativeSigner);
    return new import_EnokiKeypair.EnokiKeypair({
      address,
      ephemeralKeypair,
      maxEpoch: zkp.maxEpoch,
      proof: zkp.proof
    });
  }
}
_storageKeys = new WeakMap();
_enokiClient = new WeakMap();
_network = new WeakMap();
_encryption = new WeakMap();
_encryptionKey = new WeakMap();
_store = new WeakMap();
_idbStore = new WeakMap();
_INTERNAL_ONLY_EnokiFlow_instances = new WeakSet();
setSession_fn = async function(newValue) {
  if (newValue) {
    const storedValue = await __privateGet(this, _encryption).encrypt(
      __privateGet(this, _encryptionKey),
      JSON.stringify(newValue)
    );
    __privateGet(this, _store).set(__privateGet(this, _storageKeys).SESSION, storedValue);
  } else {
    __privateGet(this, _store).delete(__privateGet(this, _storageKeys).SESSION);
  }
  this.$zkLoginSession.set({ initialized: true, value: newValue });
};
